easybuild/tools/containers/docker.py

```diff
@@ -0,0 +1,179 @@
+#
+# https://github.com/easybuilders/easybuild
+#
+# EasyBuild is free software: you can redistribute it and/or modify
+# it under the terms of the GNU General Public License as published by
+# the Free Software Foundation v2.
+#
+# EasyBuild is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with EasyBuild.  If not, see <http://www.gnu.org/licenses/>.
+# #
+"""
+Support for generating docker container recipes and creating container images
+
+:author Mohamed Abidi (Bright Computing)
+"""
+import os
+import shutil
+import tempfile
+
+from vsc.utils import fancylogger
+
+from easybuild.framework.easyconfig.easyconfig import ActiveMNS
+from easybuild.tools.build_log import print_msg, EasyBuildError
+from easybuild.tools.config import build_option
+from easybuild.tools.config import DOCKER_BASE_IMAGE_UBUNTU, DOCKER_BASE_IMAGE_CENTOS, DEFAULT_DOCKER_BASE_IMAGE
+from easybuild.tools.run import run_cmd
+from .base import ContainerGenerator
+from .utils import det_os_deps
+
+_log = fancylogger.getLogger('tools.containers.singularity')  # pylint: disable=C0103
+
+DOCKER_UBUNTU1604_TMPL = """\
+FROM ubuntu:16.04
+LABEL maintainer=mohamed.abidi@brightcomputing.com
+
+RUN apt-get update && \\
+    apt-get install -y python python-pip lmod curl wget
+
+RUN pip install -U pip setuptools && \\
+    hash -r pip && \\
+    pip install -U easybuild
+
+RUN mkdir /app && \\
+    mkdir /scratch && \\
+    mkdir /scratch/tmp && \\
+    useradd -m -s /bin/bash easybuild && \\
+    chown easybuild:easybuild -R /app && \\
+    chown easybuild:easybuild -R /scratch
+
+RUN OS_DEPS='%(os_deps)s' && \\
+    test -n "${OS_DEPS}" && \\
+    for dep in ${OS_DEPS}; do apt-get -qq install ${dep} || true; done
+
+USER easybuild
+
+RUN set -x && \\
+    . /usr/share/lmod/lmod/init/sh && \\
+    eb %(eb_opts)s --installpath=/app/ --prefix=/scratch --tmpdir=/scratch/tmp
+
+RUN touch ${HOME}/.profile && \\
+    echo '\\n# Added by easybuild docker packaging' >> ${HOME}/.profile && \\
+    echo 'source /usr/share/lmod/lmod/init/bash' >> ${HOME}/.profile && \\
+    echo 'module use %(init_modulepath)s' >> ${HOME}/.profile && \\
+    echo 'module load %(mod_names)s' >> ${HOME}/.profile
+
+CMD ["/bin/bash", "-l"]
+"""
+
+DOCKER_CENTOS7_TMPL = """\
+FROM centos:7
+LABEL maintainer=mohamed.abidi@brightcomputing.com
+
+RUN yum install -y epel-release && \\
+    yum install -y python python-pip Lmod curl wget git
+
+RUN pip install -U pip setuptools && \\
+    hash -r pip && \\
+    pip install -U easybuild
+
+RUN OS_DEPS='%(os_deps)s' && \\
+    test -n "${OS_DEPS}" && \\
+    yum --skip-broken install -y "${OS_DEPS}" || true
+
+RUN mkdir /app && \\
```

Comment:
This last part is entirely duplicate, can we flesh this out as `DOCKER_COMMON_TMPL` and then define `DOCKER_CENTOS7_TMPL` as `"""...""" + DOCKER_COMMON_TMPL` (and the same for `DOCKER_UBUNTU1604_TMPL`)?

Or maybe use `DOCKER_CENTOS7_TMPL_HEADER` together with `DOCKER_COMMON_TMPL` in `resolve_template` below?